pkg/metrics/metrics.go

```diff
@@ -0,0 +1,239 @@
+	MachineSetInfoSpecReplicas = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mapi_machine_set_info_spec_replicas",
+		Help: "Count of the Machinesets Spec Replicas.",
+	}, []string{"name", "namespace"})
+
+	// MachineSetStatusAvailableReplicas Information of the Machinesets' status for available replicas.
+	MachineSetStatusAvailableReplicas = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mapi_machine_set_status_available_replicas",
+		Help: "Information of the mapi managed Machinesets' status for available replicas.",
+	}, []string{"name", "namespace"})
+
+	// MachineSetStatusReadyReplicas Information of the Machinesets' status for ready replicas
+	MachineSetStatusReadyReplicas = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mapi_machine_set_status_ready_replicas",
+		Help: "Information of the mapi managed Machinesets' status for ready replicas.",
+	}, []string{"name", "namespace"})
+
+	// MachineSetStatusReplicas Information of the mapi managed Machinesets' status for replicas.
+	MachineSetStatusReplicas = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mapi_machine_set_status_replicas",
+		Help: "Information of the mapi managed Machinesets' status for replicas.",
+	}, []string{"name", "namespace"})
+
+	// ScrapeFailedCounter is a Prometheus metric, which counts errors during metrics collection.
+	ScrapeFailedCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "mapi_scrape_failure_total",
+		Help: "Total count of scrape failures.",
+	}, []string{"kind", "reason"})
+)
+
+func init() {
+	prometheus.MustRegister(ScrapeFailedCounter)
+	prometheus.MustRegister(MachineInfo)
+	prometheus.MustRegister(MachineSetInfo)
+	prometheus.MustRegister(MachineSetInfoSpecReplicas)
+	prometheus.MustRegister(MachineSetStatusAvailableReplicas)
+	prometheus.MustRegister(MachineSetStatusReadyReplicas)
+	prometheus.MustRegister(MachineSetStatusReplicas)
+}
+
+type MachineCollector struct {
+	client    runtimeclient.Client
+	namespace string
+}
+
+func getClient(kubeconfig string) (runtimeclient.Client, error) {
+	var config *rest.Config
+	if kubeconfig != "" {
+		c, err := clientcmd.LoadFromFile(kubeconfig)
+		if err != nil {
+			return nil, fmt.Errorf("error loading KubeConfig: %v", err.Error())
+		}
+		config, err = clientcmd.NewDefaultClientConfig(*c, &clientcmd.ConfigOverrides{}).ClientConfig()
+		if err != nil {
+			return nil, fmt.Errorf("error loading KubeConfig: %v", err.Error())
+		}
+	} else {
+		glog.V(4).Infof("Using in-cluster kube client config")
+		var err error
+		config, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, fmt.Errorf("error loading in-cluster Kube client Config: %v", err.Error())
+		}
+	}
+	client, err := runtimeclient.New(config, runtimeclient.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("error creating runtime client: %v", err.Error())
+	}
+
+	return client, nil
+}
+
+func NewMachineCollector(kubeconfig string) (*MachineCollector, error) {
```

Comment:
Don't we construct a client right after initializing this when the operator starts? Would it be easier to just pass a client instead of a kubeconfig?

Reply:
mao creates two clients, one for concrete kubernetes apis and other for openshift specific apis. These clients will not work for CRDs. I think, in future we can try using controller-runtime abstractions/client instead of using client-go library directly for creating k8s client and os client. And then, may be able to work with a single controller-runtime provided client.

Reply:
`These clients will not work for CRDs` not sure what you mean, we could analogous create another client for machineAPI or even better use listers() as we do for the deployments

Reply:
`dynamic` and `kubernetes` are two packages of client-go. `kubernetes` package provides clientset and Kubernetes resource specific clientset methods for accessing standard Kubernetes APIs.  Deployments are standard kubernetes api type. Informers are then built on top of this client to use cache.
mao is using `kubernetes` pkg. [kubeclient](https://github.com/openshift/machine-api-operator/blob/master/cmd/machine-api-operator/controller_context.go#L30) returns [interface](https://github.com/kubernetes/client-go/blob/master/kubernetes/clientset.go#L63) which has only standard k8s apis. 
For CRDs, [dynamic](https://github.com/kubernetes/client-go/tree/master/dynamic) client is to be used. 

controller-runtime provides [ a single abstracted client over](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client.go#L46) these two type of clients that client-go supports. This single client automatically handles both type of queries. I am using this controller-runtime provided client here to list machines. It already using informers underneath.

If we want to use a single client, we might have to make mao to use controller-runtime provided client and then same could be used by metrics pkg.

Please correct if something i misunderstood.

Reply:
you can do same we do for featuregates/deployments, see https://github.com/openshift/cluster-api/tree/openshift-4.2-cluster-api-0.1.0/pkg/client
"It already using informers underneath." that's true when instantiated via manager, not in this case
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L261
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L278